File: 5 semestre/Programacao para Dispositivos Moveis/UAS/UA1/java/src/empresa/Empresa.java
package empresa;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

class Empresa {
    private String nome;
    private String cnpj;
    private List<Funcionario> funcionarios;
    private FolhaPagamento folha;

    public Empresa(String nome, String cnpj) {
        this.nome = nome;
        this.cnpj = cnpj;
        this.funcionarios = new ArrayList<>();
        this.folha = new FolhaPagamento();
    }

    public void contratar(Funcionario f) {
        if (f != null && !funcionarios.contains(f)) {
            funcionarios.add(f);
        }
    }

    public void demitir(String id) {
        funcionarios.stream().filter(f -> f.getId().equals(id)).findFirst().ifPresent(f -> {
            f.desativar();
            System.out.println("Funcionário " + f.getNome() + " desligado.");
        });
    }

    public List<Funcionario> getFuncionariosAtivos() {
        return funcionarios.stream().filter(Funcionario::isAtivo).collect(Collectors.toList());
    }

    public List<Funcionario> getFuncionariosPorDepartamento(Departamento d) {
        return funcionarios.stream().filter(f -> f.getDepartamento() == d && f.isAtivo()).collect(Collectors.toList());
    }

    public List<Desenvolvedor> getDesenvolvedoresPorLinguagem(String linguagem) {
        return funcionarios.stream()
                .filter(f -> f instanceof Desenvolvedor && f.isAtivo())
                .map(f -> (Desenvolvedor) f)
                .filter(dev -> dev.getLinguagens().contains(linguagem))
                .collect(Collectors.toList());
    }

    public Optional<Funcionario> buscarFuncionarioPorCpf(String cpf) {
        return funcionarios.stream().filter(f -> f.getCpf().equals(cpf)).findFirst();
    }

    public void aplicarAumentoGlobal(double percentual) {
        for (Funcionario f : funcionarios) {
            if (f.isAtivo()) {
                f.aumentarSalario(percentual);
            }
        }
    }

    public double calcularTotalFolha() {
        return funcionarios.stream().filter(Funcionario::isAtivo).mapToDouble(Funcionario::calcularSalarioTotal).sum();
    }

    public void listarFuncionariosOrdenadosPorSalario() {
        funcionarios.stream()
                .filter(Funcionario::isAtivo)
                .sorted(Comparator.comparingDouble(Funcionario::calcularSalarioTotal).reversed())
                .forEach(System.out::println);
    }

    public void processarPagamentoTodos() {
        System.out.println("=== PROCESSANDO FOLHA ===");
        for (Funcionario f : getFuncionariosAtivos()) {
            try {
                folha.processarPagamento(f);
            } 
            
            catch (FuncionarioInativoException e) {
                System.err.println(e.getMessage());
            }
        }
        System.out.println("Total pago: R$ " + String.format("%.2f", folha.totalPago()));
    }

    public void promoverDesenvolvedor(String id, NivelExperiencia novoNivel) {
        funcionarios.stream()
                .filter(f -> f.getId().equals(id) && f instanceof Desenvolvedor && f.isAtivo())
                .map(f -> (Desenvolvedor) f)
                .findFirst()
                .ifPresent(dev -> {
                    dev.aumentarSalario(15);
                    System.out.println("Desenvolvedor " + dev.getNome() + " promovido de " + dev.getNivel() + " para " + novoNivel);
                });
    }
}

File: 5 semestre/Programacao para Dispositivos Moveis/UAS/UA1/dart/empresa/empresa.dart
class Empresa {
  String nome;
  String cnpj;
  List<Funcionario> funcionarios;
  FolhaPagamento folha;

  Empresa(this.nome, this.cnpj)
      : funcionarios = [],
        folha = FolhaPagamento();

  void contratar(Funcionario f) {
    if (f != null && !funcionarios.contains(f)) {
      funcionarios.add(f);
    }
  }

  void demitir(String id) {
    Funcionario? func = funcionarios.cast<Funcionario?>().firstWhere(
        (f) => f!.id == id,
        orElse: () => null);
    if (func != null) {
      func.desativar();
      print("Funcionário ${func.nome} desligado.");
    }
  }

  List<Funcionario> getFuncionariosAtivos() {
    return funcionarios.where((f) => f.ativo).toList();
  }

  List<Funcionario> getFuncionariosPorDepartamento(Departamento d) {
    return funcionarios.where((f) => f.departamento == d && f.ativo).toList();
  }

  List<Desenvolvedor> getDesenvolvedoresPorLinguagem(String linguagem) {
    return funcionarios
        .where((f) => f is Desenvolvedor && f.ativo)
        .map((f) => f as Desenvolvedor)
        .where((dev) => dev.linguagens.contains(linguagem))
        .toList();
  }

  Funcionario? buscarFuncionarioPorCpf(String cpf) {
    try {
      return funcionarios.firstWhere((f) => f.cpf == cpf);
    } catch (_) {
      return null;
    }
  }

  void aplicarAumentoGlobal(double percentual) {
    for (var f in funcionarios) {
      if (f.ativo) {
        f.aumentarSalario(percentual);
      }
    }
  }

  double calcularTotalFolha() {
    return funcionarios
        .where((f) => f.ativo)
        .fold(0.0, (sum, f) => sum + f.calcularSalarioTotal());
  }

  void listarFuncionariosOrdenadosPorSalario() {
    var ativos = getFuncionariosAtivos();
    ativos.sort((a, b) => b.calcularSalarioTotal().compareTo(a.calcularSalarioTotal()));
    for (var f in ativos) {
      print(f);
    }
  }

  void processarPagamentoTodos() {
    print("=== PROCESSANDO FOLHA ===");
    for (var f in getFuncionariosAtivos()) {
      try {
        folha.processarPagamento(f);
      } on FuncionarioInativoException catch (e) {
        print(e);
      }
    }
    print("Total pago: R\$ ${folha.totalPago().toStringAsFixed(2)}");
  }

  void promoverDesenvolvedor(String id, NivelExperiencia novoNivel) {
    Funcionario? func = funcionarios.firstWhere(
        (f) => f.id == id && f is Desenvolvedor && f.ativo,
        orElse: () => null);
    if (func != null) {
      Desenvolvedor dev = func as Desenvolvedor;
      dev.aumentarSalario(15);
      print(
          "Desenvolvedor ${dev.nome} promovido de ${dev.nivel} para $novoNivel");
    }
  }
}